### PR TITLE
Feature/mktmpdir

### PR DESCRIPTION
**Pull request by bramswenson**
This is simply a rebase of #2911 to current master (03-01-2017)

### CODE REVIEW

**Comment by marceloboeira** [comment]
👍

File: src/dir.cr [hunk 1]
```diff
@@ -215,6 +215,56 @@ class Dir
...
+    fileno = LibC.mkdtemp(tmp_dir)
+    if fileno == nil
+      raise Errno.new("mkdtemp")
+    end
+
+    tmp_dir
+  end
+
+  # Creates a new temporary directory within the lifecycle
+  # of the given block and destroys it, and its content, later on.
```
Comment (RX14) [inline src/dir.cr]: `...and its content, when the block returns`

File: src/dir.cr [hunk 2]
```diff
@@ -215,6 +215,56 @@ class Dir
...
+    tmp_dir
+  end
+
+  # Returns the tmp dir
+  #
+  # ```
+  # Dir.tmpdir # => "/tmp"
+  # ```
+  def self.tmpdir
+    unless tmpdir = ENV["TMPDIR"]?
```
Comment (RX14) [inline src/dir.cr]: This is a bit confusing, how about 

```cr
tmpdir = ENV["TMPDIR"]? || "/tmp"
```

File: src/dir.cr (hunk 1 above)
Comment (marceloboeira) [inline src/dir.cr]: 👍

File: src/dir.cr (hunk 2 above)
Comment (marceloboeira) [inline src/dir.cr]: 👍

**Comment by marceloboeira** [comment]
@bramswenson do you want to make the changes? because if I would require access to your fork/branch.

**Comment by bramswenson** [comment]
@marceloboeira @RX14 Done! Thanks!

**Comment by marceloboeira** [comment]
@bramswenson great!

File: src/dir.cr [hunk 3]
```diff
@@ -223,6 +271,28 @@ class Dir
     0
   end
 
+  # Deletes a file or directory *path*
+  # If *path* is a directory, this method removes all its contents recursively
+  # ```
+  # Dir.rm_r("dir")
+  # Dir.rm_r("file.cr")
+  # ```
+  def self.rm_r(path : String)
```
Comment (asterite) [inline src/dir.cr]: This belongs to FileUtils, not Dir

**Comment by asterite** [comment]
@bramswenson Thank you for this!

In general the PR looks very good. However:
* I'm not sure about adding `Dir.rm_r`. In general we have simple methods in File and Dir, and more complex methods in FileUtils. FileUtils has methods that you can find in bash, like `rm_r` or `mkdir_p`.
* I'm also not sure about `Dir.tmpdir`, but I think it's OK (similar to Ruby, related to Dir and not Tempfile), but this is a small breaking change and we should note it.

I'm all the time unsure about File, Dir, FileUtils and where to put things. Maybe things are too complex/messy right now. I'll discuss this with the team, maybe we can just put all methods in File and Dir and remove FileUtils.

**Comment by asterite** [comment]
For now we can merge your change if it doesn't introduce breaking changes (Use `Tmpfile.dirname` instead of `Dir.tmpdir` and keeps `FileUtils.rm_r` instead of `Dir.rm_r`)). That's at least if we want this included in the next patch release.

**Comment by bramswenson** [comment]
@asterite Thanks for the kind words, but I did nothing but the rebase to master out of selfish desire to use this feature ;) @marceloboeira Did the actual work and there was reasonable discussion about the move between `Dir` and `FileUtils` in #2911 between @marceloboeira and @RX14. I really don't have a dog in that fight and will be happy to see the feature wherever you all decide is most appropriate. Thanks!

**Comment by bramswenson** [comment]
I will say that the removal of `FileUtils` in preference for `File` and `Dir` makes sense to me. I also think these methods should closely resemble their unix counterparts in naming as unix file modification executable names are common knowledge (or should be). But that is just my 2 cents.

**Comment by marceloboeira** [comment]
I didn't have time to take a look at the comments yet folks, but I will do it soon. @bramswenson thanks for the invitation to your repository and the efforts on making it happen. 👍

File: src/dir.cr (hunk 3 above)
Comment (sevk) [inline src/dir.cr]: Dir can rm_r , FileUtils can rm_r too .
 just like String can have add method , Number can add  too . 
Right ?

Comment (Sija) [inline src/dir.cr]: @sevk stop trolling please.

**Comment by bew** [comment]
I was about to (suggest to) add this!
@bramswenson, will you work on this anytime soon ? (I'd really like something like that to refactor some compiler specs)

About naming I was thinking, since we already have the `Tempfile` class, we could have `Tempdir` ?
And put the two in a `Temporary` module with a simple API like:
```cr
Temporary::File.new # gives a file
Temporary::Dir.new # gives a dir

Temporary::Dir.open do |dir|
  # do something with the `dir`
  puts dir.path
end # the block ensures that the temp dir is deleted

Temporary.create_name("foo.XXX") # gives a (file/dir)name following a pattern
# etc..
```

**Comment by RX14** [comment]
I think we shouldn't have "tempdir" and "tempfile" classes, just methods on `Dir` and `File` which create tempoary versions of themselves.

**Comment by bew** [comment]
> Why does `Tempfile` currently not inherit `File`? It makes little sense to me.

I thought the same, I'll fix that.

**Comment by marceloboeira** [comment]
@RX14 any updates here?

**Comment by RX14** [comment]
@marceloboeira `Tempfile` was removed in #6485. We still need to decide what to do about `FileUtils`, I say that most of it's methods belong on `Path` once that is merged.

**Comment by j8r** [comment]
Too old, I think should be closed now. A new PR can be opened if wanted.

As now, to create a temp dir:
```cr
dir = File.tempname 
Dir.mkdir dir
begin
  # do things
ensure
  FileUtils.rm_rf dir
end
```

**Pull request closed without merging**